servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -456,25 +378,48 @@ public void completeFile(long fileId)
         }
       }
 
-      completeFileInternal(fileInode.getBlockIds(), fileId, fileLength, false, opTimeMs);
+      // If the file is both persisted and stored in memory, the memory footprint and the UFS file
+      // length should match.
+      if (fileInode.isPersisted() && inMemoryLength != 0
+          && options.getUfsLength() != inMemoryLength) {
+        throw new SuspectedFileSizeException("Inconsistent file length: Tachyon " + inMemoryLength
+            + " UFS " + options.getUfsLength());
+      }
+
+      // If the file is persisted, its length is determined by UFS. Otherwise, its length is
+      // determined by its memory footprint.
+      long length = fileInode.isPersisted() ? options.getUfsLength() : inMemoryLength;
+
+      completeFileInternal(fileInode.getBlockIds(), fileId, length, opTimeMs);
       writeJournalEntry(
-          new CompleteFileEntry(fileInode.getBlockIds(), fileId, fileLength, opTimeMs));
+          new CompleteFileEntry(fileInode.getBlockIds(), fileId, length, opTimeMs));
       flushJournal();
     }
   }
 
-  void completeFileInternal(List<Long> blockIds, long fileId, long fileLength, boolean replayed,
-      long opTimeMs) throws FileDoesNotExistException, InvalidPathException {
+  void completeFileInternal(List<Long> blockIds, long fileId, long length, long opTimeMs)
+      throws FileDoesNotExistException, InvalidPathException, SuspectedFileSizeException {
     // This function should only be called from within synchronized (mInodeTree) blocks.
-    InodeFile inodeFile = (InodeFile) mInodeTree.getInodeById(fileId);
-    inodeFile.setBlockIds(blockIds);
-    inodeFile.setCompleted(fileLength);
-    inodeFile.setLastModificationTimeMs(opTimeMs);
+    InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+    inode.setBlockIds(blockIds);
+    inode.setLastModificationTimeMs(opTimeMs);
+    inode.complete(length);
+
+    if (inode.isPersisted()) {
+      // Commit all the file blocks (without locations) so the metadata for the block exists.
+      long currLength = length;
+      for (long blockId : inode.getBlockIds()) {
+        long blockSize = Math.min(currLength, inode.getBlockSizeBytes());
```

Comment:
since this is only used for marking the block as committed in the UFS, I think it is fine